### PR TITLE
Add instance tags [run-systemtest]

### PR DESCRIPTION
**Pull request by bratseth**
@jonmv please review, but we can hold off merge until after tomorrows arch meeting

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/pkg/ApplicationPackage.java [hunk 1]
```diff
@@ -202,7 +203,8 @@ private void preProcessAndPopulateCache() {
                                     new InputStreamReader(new ByteArrayInputStream(servicesXml.content()), UTF_8),
                                     InstanceName.defaultName(),
                                     Environment.prod,
-                                    RegionName.defaultName())
+                                    RegionName.defaultName(),
+                                    Tags.empty())
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/pkg/ApplicationPackage.java:207]: Urgh, I guess this should preprocess with all configurations that may eventually be deployed.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 2]
```diff
@@ -2035,7 +2036,7 @@ private HttpResponse createInstance(String tenantName, String applicationName, S
         if (controller.applications().getApplication(applicationId).isEmpty())
             createApplication(tenantName, applicationName, request);
 
-        controller.applications().createInstance(applicationId.instance(instanceName));
+        controller.applications().createInstance(applicationId.instance(instanceName), Tags.empty());
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java:2039]: Hmm ... supporting `dev/perf` here isn't trivial, and perhaps not that useful?

File: configserver/src/main/java/com/yahoo/vespa/config/server/session/PrepareParams.java [hunk 3]
```diff
@@ -273,6 +306,7 @@ public static PrepareParams fromHttpRequest(HttpRequest request, TenantName tena
                             .verbose(request.getBooleanProperty(VERBOSE_PARAM_NAME))
                             .timeoutBudget(SessionHandler.getTimeoutBudget(request, barrierTimeout))
                             .applicationId(createApplicationId(request, tenant))
+                            .tags(Tags.fromString(request.getProperty(TAGS_PARAM_NAME)))
```
Comment (jonmv) [inline configserver/src/main/java/com/yahoo/vespa/config/server/session/PrepareParams.java:309]: This PR doesn't set this value.

**Review by jonmv — APPROVED** [review]
Great! So much wiring 😝

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/pkg/ApplicationPackage.java (hunk 1 above)
Comment (bratseth) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/pkg/ApplicationPackage.java:207]: I don't know what this is for, hopefully it is just wasted work if there are variants?

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java (hunk 2 above)
Comment (bratseth) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java:2039]: I assumed not.

File: configserver/src/main/java/com/yahoo/vespa/config/server/session/PrepareParams.java (hunk 3 above)
Comment (bratseth) [reply configserver/src/main/java/com/yahoo/vespa/config/server/session/PrepareParams.java:309]: How do you figure that? Afaik this payload is autogenerated from com.yahoo.vespa.hosted.controller.api.application.v4.model.DeploymentData, which sets tags on line 655 in ApplicationController.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/pkg/ApplicationPackage.java (hunk 1 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/pkg/ApplicationPackage.java:207]: It is for storing all relevant deployment meta data to our audit store. If a certain override includes a file which isn't included in the default path taken here, that file will not be stored. It's not critical. Perhaps we could instead run only the include processor, and not the override one? Would that include all possible variants? (We don't care about the generated XML, but all files read by the inclusion process will be loaded to the file cache, which is then flushed to the audit storage.)

Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/pkg/ApplicationPackage.java:207]: You don't need to fix this in this PR.

File: configserver/src/main/java/com/yahoo/vespa/config/server/session/PrepareParams.java (hunk 3 above)
Comment (jonmv) [reply configserver/src/main/java/com/yahoo/vespa/config/server/session/PrepareParams.java:309]: The `DeploymentData` has the tags, but they're not read anywhere. The implementation of the `ConfigServer` needs to do that.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/pkg/ApplicationPackage.java (hunk 1 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/pkg/ApplicationPackage.java:207]: Ah, never mind, the include processor is run first.

File: configserver/src/main/java/com/yahoo/vespa/config/server/session/PrepareParams.java (hunk 3 above)
Comment (jonmv) [reply configserver/src/main/java/com/yahoo/vespa/config/server/session/PrepareParams.java:309]: Sorry, this wasn't the relevant one. The `fromJson` is it. I'll make a PR for you :)

Comment (bratseth) [reply configserver/src/main/java/com/yahoo/vespa/config/server/session/PrepareParams.java:309]: The fromJson does reads the tags already?

Comment (jonmv) [reply configserver/src/main/java/com/yahoo/vespa/config/server/session/PrepareParams.java:309]: Yes, but they're not written anywhere.

File: config-provisioning/src/test/java/com/yahoo/config/provision/TagsTest.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+ * @author bratseth
+ */
+public class TagsTest {
+
+    @Test
+    public void testEmpty() {
+        assertEquals(Tags.empty(), Tags.fromString(null));
+        assertEquals(Tags.empty(), Tags.fromString(""));
+        assertEquals(Tags.empty(), Tags.fromString(" "));
+    }
```
Comment (jonmv) [inline config-provisioning/src/test/java/com/yahoo/config/provision/TagsTest.java:22]: How about this?

```java
        assertEquals(new Tags(Set.of("tag1", "tag2")), Tags.fromString("  tag1     tag2  "));
```

Comment (jonmv) [reply config-provisioning/src/test/java/com/yahoo/config/provision/TagsTest.java:22]: This currently fails with duplicate element (`""`).

Comment (jonmv) [reply config-provisioning/src/test/java/com/yahoo/config/provision/TagsTest.java:22]:
```suggestion

    @Test
    public void testDeserialization() {
        assertEquals(new Tags(Set.of("tag1", "tag2")), Tags.fromString("  tag1     tag2  "));
    }
```

**Comment by bratseth** [comment]
Thanks!

**Comment by jonmv** [comment]
Also need a stable ordering for tests ...

**Pull request merged**